### PR TITLE
Add bound checks for each dimension of `mdspan`

### PR DESCRIPTION
**Pull request by fbusato**
Fixes https://github.com/NVIDIA/cccl/issues/3063 and https://github.com/NVIDIA/cccl/issues/3045

## Description

Add bound checks for each dimension of `mdspan`. This replaces the global index check

### CODE REVIEW

**Comment by fbusato** [comment]
The code looks pretty bad because we don't have folding expression in c++14 and it needs to work with zero-sizes `mdspan

File: libcudacxx/include/cuda/std/__mdspan/mdspan.h [hunk 1]
```diff
@@ -108,28 +110,45 @@ class mdspan
...
     }
+
+    template <class... _SizeTypes>
+    _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr bool
+    __check_index(_Extents const& exts, _SizeTypes... __indices)
+    {
+      index_type __zero = 0; // avoid warning:186 (ICC/NVCC) pointless comparison of unsigned integer with zero
+      array<bool, sizeof...(_SizeTypes)> __res{((is_signed_v<index_type> && static_cast<index_type>(__indices) < __zero)
+                                                || static_cast<index_type>(__indices) >= exts.extent(_Idxs))...};
+      _LIBCUDACXX_UNUSED_VAR(__zero * 2);
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__mdspan/mdspan.h:121]: I believe we should directly check against `npos` rather than < 0

File: libcudacxx/include/cuda/std/__mdspan/mdspan.h [hunk 2]
```diff
@@ -108,28 +110,45 @@ class mdspan
...
+      return _CUDA_VSTD::all_of(__res.begin(), __res.end(), [](bool __v) {
+        return __v;
+      });
+    }
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__mdspan/mdspan.h:125]: This will instantiate one lambda per extents. We can simply pass cuda::std::identity

**Review by miscco — CHANGES_REQUESTED** [review]

**Comment by fbusato** [comment]
nice to see that by adding strict bound checks discovered 8 tests failing

File: libcudacxx/include/cuda/std/__mdspan/mdspan.h (hunk 1 above)
Comment (fbusato) [reply libcudacxx/include/cuda/std/__mdspan/mdspan.h:121]: `npos` is for strings and maps to `-1`, while for `mdspan` we need to check against `dynamic_extent` which maps to the maximum value https://en.cppreference.com/w/cpp/container/span/dynamic_extent

**Comment by miscco** [comment]
Make sure that those tests are actually failing and not some corner cases for e.g submdspan and `required_span_size()`

**Review by bernhardmgruber — APPROVED** [review]

File: libcudacxx/include/cuda/std/__mdspan/mdspan.h [hunk 3]
```diff
@@ -108,28 +113,44 @@ class mdspan
...
+    __check_index(_Extents const& exts, _SizeTypes... __indices)
+    {
+      // std::array supports zero size
+      // ternary avoids "pointless comparison of unsigned integer with zero" warning
+      array<bool, sizeof...(_SizeTypes)> __res{
+        ((is_unsigned_v<index_type> ? true : static_cast<index_type>(__indices) >= 0)
+         && static_cast<index_type>(__indices) < exts.extent(_Idxs))...};
+      return _CUDA_VSTD::all_of(__res.begin(), __res.end(), identity{});
+    }
+
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__mdspan/mdspan.h:128]: Thinking about this a bit more I would love to have an overload that separately checks the zero rank case 
```suggestion
    _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr bool
    __check_index(_Extents const& exts) const {
      static_assert(_Extents::rank() == 0, "Rank must be zero for access without indices");
      return true;
    }

    template <class... _SizeTypes>
    _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr bool
    __check_index(_Extents const& exts, _SizeTypes... __indices)
    {
      // ternary avoids "pointless comparison of unsigned integer with zero" warning
      bool __res[ sizeof...(_SizeTypes)]{
        ((is_unsigned_v<index_type> ? true : static_cast<index_type>(__indices) >= 0)
         && static_cast<index_type>(__indices) < exts.extent(_Idxs))...};
      return _CUDA_VSTD::all_of(__res.begin(), __res.end(), identity{});
    }

```

Comment (bernhardmgruber) [reply libcudacxx/include/cuda/std/__mdspan/mdspan.h:128]: We may be able to just fold in a few weeks of time:

```suggestion

    template <class... _SizeTypes>
    _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr bool
    __check_index(_Extents const& exts, _SizeTypes... __indices)
    {
      return
        (((is_unsigned_v<index_type> ? true : static_cast<index_type>(__indices) >= 0)
         && static_cast<index_type>(__indices) < exts.extent(_Idxs))) && ...);
    }

```
which is my preferred solution.

Comment (fbusato) [reply libcudacxx/include/cuda/std/__mdspan/mdspan.h:128]: Honestly, I would prefer the later one. It is much less code, more readable and easy to understand.

**Review by bernhardmgruber — APPROVED** [review]

**Review by miscco — APPROVED** [review]
